manager/allocator/network/port/doc.go

```diff
@@ -0,0 +1,33 @@
+package port
+
+// The `port` package contains the implementation of the swarmkit Allocator for
+// network ports. Its purpose is to assign ports to Endpoint objects based on
+// the provided Spec. For ingress ports, it keeps track of the availability of
+// ports, and handles dynamic allocation of ports when the user has specified
+// no PublishedPort. The allocator does _not_ keep track of host-mode ports in
+// any capacity. It silently adds them to the returned Ports assignment,
+// without checking their validity or availability. Port assignments for
+// host-mode ports are handled on the worker.
```

Comment:
Also, what about this case:
```
docker service create --publish published=30000,target=80,mode=host nginx
docker service create --publish published=0,target=80               nginx
```
Doesn't the allocator need to know that a host is using port 30000?

(I just tried it with 18.03.1-ce and it let me use the same port for two services. Requests always seemed to go to the ingress service until I killed that service, then they went to the host service.)

Reply:
the scheduler is not involved in host mode ports. it was a very rushed feature, if i recall correctly, and it's sensitive to collisions.

the problem with handling host-mode ports at the cluster level is that it involves an interaction between the allocator and the scheduler. the scheduler won't know until the task passes through the allocator what ports it's using, but the allocator won't know until it passes through the scheduler what node it's on. plus, you can dynamically assign ports, which means you really actually can't know what port will be chosen on the target node.

so, the solution now is to pass host ports as-is, and let the worker fail the task if there is a collision with a local port (thus pushing the scheduler to reschedule the task, hopefully on a different node with the requested port available).

Reply:
But if a host port isn't available on a node because it's already used as an ingress port, won't we have the same problem on all other nodes of the cluster too? Should a worker reject a request to forward an ingress port if it is already using it as a host-mode port? (I'm not too clear about how overlay networks work currently)

Reply:
This interaction between the scheduler and the allocator has been discussed in other Issues and PRs. It has been suggested to extract the allocators from the pipeline:

Current:
```
                    MANAGER NODE (LEADER)                           WORKER NODE
_______________ ______________ ______________ ______________     _______________
\              \              \              \              \    \              \
 | Orchestrator |  Allocator   |  Scheduler   |  Dispatcher  |    |    Agent     |
/______________/______________/______________/______________/    /______________/
```

Proposed:
```
             MANAGER NODE (LEADER)                   WORKER NODE
_______________ ______________ ______________     _______________
\              \              \              \    \              \
 | Orchestrator |  Scheduler   |  Dispatcher  |    |    Agent     |
/______________/______________/______________/    /______________/
                       |
                 +-----+-----+
                 | Allocator |
                 +-----------+
```
This way the port allocator could be called before (or at the start of) the scheduler but other allocators could be called later (resources could be consider allocatable items for example).